cmd/runtimetest/main.go

```diff
@@ -410,201 +578,379 @@ func validateDefaultFS(spec *rspec.Spec, t *tap.T) error {
+	}
+	expectedType := device.Type
+	if expectedType == "u" {
+		expectedType = "c"
+	}
+	var devType string
+	switch fStat.Mode & syscall.S_IFMT {
+	case syscall.S_IFCHR:
+		devType = "c"
+	case syscall.S_IFBLK:
+		devType = "b"
+	case syscall.S_IFIFO:
+		devType = "p"
+	default:
+		devType = "unmatched"
+	}
+	rfcError, err = c.Ok(devType == expectedType, specerror.DevicesAvailable, version, fmt.Sprintf("%s has the expected type", description))
+	if err != nil {
+		return err
+	}
+	c.harness.YAML(map[string]string{
+		"level":     rfcError.Level.String(),
+		"reference": rfcError.Reference,
+		"path":      device.Path,
+		"expected":  expectedType,
+		"actual":    devType,
+	})
+	if devType != expectedType {
+		return nil
+	}
+
+	if devType != "p" {
+		dev := fStat.Rdev
+		major := (dev >> 8) & 0xfff
+		minor := (dev & 0xff) | ((dev >> 12) & 0xfff00)
+		rfcError, err = c.Ok(int64(major) == device.Major, specerror.DevicesAvailable, version, fmt.Sprintf("%s has the expected major ID", description))
+		if err != nil {
+			return err
 		}
-		if device.FileMode != nil {
-			expectedPerm := *device.FileMode & os.ModePerm
-			actualPerm := fi.Mode() & os.ModePerm
-			if expectedPerm != actualPerm {
-				return fmt.Errorf("%v filemode expected is %v, actual is %v", device.Path, expectedPerm, actualPerm)
-			}
+		c.harness.YAML(map[string]interface{}{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"path":      device.Path,
+			"expected":  device.Major,
+			"actual":    major,
+		})
+		rfcError, err = c.Ok(int64(minor) == device.Minor, specerror.DevicesAvailable, version, fmt.Sprintf("%s has the expected minor ID", description))
+		if err != nil {
+			return err
 		}
-		if device.UID != nil {
-			if *device.UID != fStat.Uid {
-				return fmt.Errorf("%v uid expected is %v, actual is %v", device.Path, *device.UID, fStat.Uid)
-			}
+		c.harness.YAML(map[string]interface{}{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"path":      device.Path,
+			"expected":  device.Minor,
+			"actual":    minor,
+		})
+	}
+
+	if device.FileMode == nil {
+		c.harness.Skip(1, fmt.Sprintf("%s has unconfigured permissions", description))
+	} else {
+		expectedPerm := *device.FileMode & os.ModePerm
+		actualPerm := fi.Mode() & os.ModePerm
+		rfcError, err = c.Ok(actualPerm == expectedPerm, specerror.DevicesAvailable, version, fmt.Sprintf("%s has the expected permissions", description))
+		if err != nil {
+			return err
 		}
-		if device.GID != nil {
-			if *device.GID != fStat.Gid {
-				return fmt.Errorf("%v uid expected is %v, actual is %v", device.Path, *device.GID, fStat.Gid)
-			}
+		c.harness.YAML(map[string]interface{}{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"path":      device.Path,
+			"expected":  expectedPerm,
+			"actual":    actualPerm,
+		})
+	}
+
+	if description == "/dev/console (default device)" {
```

Comment:
wouldn't it be better to test
```
if device.Path == "/dev/console" {
```

Reply:
> wouldn't it be better to test
>
>    if device.Path == "/dev/console" {

I'm fine with that.  My current check ensures that we *do* perform the rest of these checks if the config JSON explicitly included `/dev/console` in `linux.devices`.  But we don't do that in `validation/`, and I don't expect us to start, so it probably doesn't matter either way.